qiskit/dagcircuit/dagcircuit.py

```diff
@@ -1112,93 +1018,157 @@ def replace_block_with_op(self, node_block, op, wire_pos_map, cycle_check=True):
         for nd in node_block:
             self._decrement_op(nd.op)
 
-    def substitute_node_with_dag(self, node, input_dag, wires=None):
+    def _substitute_node_with_dag__wire_map(self, node, input_dag, wires, propagate_condition):
+        """Build the wire mappings for :meth:`.substitute_node_with_dag, returning a mapping of
+        wires in ``input_dag`` to wires in ``self``.  This is the canonicalisation routine for the
+        ``wires`` argument of :meth:`.DAGCircuit.substitute_node_with_dag`."""
+        if isinstance(wires, dict):
+            wire_map = wires
+        else:
+            wires = input_dag.wires if wires is None else wires
+            node_cargs = set(node.cargs)
+            node_wire_order = list(node.qargs) + list(node.cargs)
+            # If we're not propagating it, the number of wires in the input DAG should include the
+            # condition as well.
+            if not propagate_condition:
+                node_wire_order += [
+                    bit
+                    for bit in self._bits_in_condition(getattr(node.op, "condition", None))
+                    if bit not in node_cargs
+                ]
+            if len(wires) != len(node_wire_order):
+                raise DAGCircuitError(f"expected {len(node_wire_order)} wires, got {len(wires)}")
+            wire_map = dict(zip(wires, node_wire_order))
+            if len(wire_map) != len(node_wire_order):
+                raise DAGCircuitError("duplicate wires")
+        for input_dag_wire, our_wire in wire_map.items():
+            if our_wire not in self.input_map:
+                raise DAGCircuitError(f"{our_wire} is not in this DAG")
```

Comment:
Yeah, I'm not a big fan, and they just suck up CPU mostly unnecessarily.  I'm happy to remove them if you're on board too.  That would also mean that this function would be short enough to inline back into the main body.  If that's ok by you, I can probably refactor the other two helper methods into private methods outside of `DAGCircuit` (and we can potentially follow up to swap the other use of `DAGCircuit._map_condition` in `compose` over to the new form).

Reply:
To the best of my knowledge, (at least some of) these date back to the initial `DAGCircuit` implementation.

I do think they have a use, even though they're not user-actionable errors, in that they guard against conditions which can lead to invalid or incorrect output circuits (where we drop gates, conditions, etc.). Ideally, these are conditions that shouldn't ever happen (but in practice still sometimes do). These serve as guardrails either when we're refactoring passes or when users are working with the `DAGCircuit` to say "you're off the well beaten path, there's a bug somewhere, etc."

I'd be in favor of removing them if we can find an alternate way of catching these cases (through restructuring APIs, adding more thorough DAG validation that can be used in testing, or otherwise), but against dropping them outright.

Reply:
Well-written tests should already catch all these errors, I think, because passes should be comparing their outputs to expected circuits.  Equally, though, that means that users shouldn't be seeing these exceptions being generated other than by their own code calling `DGCircuit.substitute_node_with_dag` directly, so the concern about them being unactionable to users might not be so important.

I've left the error-checking in place, but inlined the helper function back into `substitute_node_with_dag` and added a "bit mapping invalid:" prefix to the messages to hopefully make the tracebacks and errors a little clearer, in 110757fa9.  Happy to do a bit more, if we'd prefer.

Reply:
Yeah, I guess it comes down to what are usage expectations are for the API. For internal terra passes I agree we're unlikely to ever see this because we have sufficient testing. But it is possible (and definitely more likely) that an out of tree pass author could use the wire mapping argument incorrectly. That was really what I was thinking about for user actionability because if a user installs a bad/buggy plugin and trips these conditions I'm not sure they'd know what to do with these exceptions so low in the stack. But I guess I just talked myself into keeping the checks for that use case and the extra context you added to the error message is probably sufficient.